service/history/queues/executable.go

```diff
@@ -334,62 +312,112 @@ func (e *executableImpl) HandleErr(err error) (retErr error) {
 			}
 			e.resourceExhaustedCount++
 			metrics.TaskThrottledCounter.With(e.taggedMetricsHandler).Record(1)
-			return err
+			return true, err
 		}
 
 		err = consts.ErrResourceExhaustedBusyWorkflow
 	}
 	e.resourceExhaustedCount = 0
 
 	if _, isNotFound := err.(*serviceerror.NotFound); isNotFound {
-		return nil
+		return true, nil
 	}
 
 	// This means that namespace is deleted, and it is safe to drop the task (=ignore the error).
 	if _, isNotFound := err.(*serviceerror.NamespaceNotFound); isNotFound {
-		return nil
+		return true, nil
+	}
+
+	if _, isDataLoss := err.(*serviceerror.DataLoss); isDataLoss {
+		e.logger.Error("Dropping task due to data loss", tag.Error(err))
+		return true, nil
 	}
 
+	if _, ok := err.(*serviceerror.NamespaceNotActive); ok {
+		// error is expected when there's namespace failover,
+		// so don't count it into task failures.
+		metrics.TaskNotActiveCounter.With(e.taggedMetricsHandler).Record(1)
+		return true, err
+	}
+
+	return false, err
+}
+
+// Terminal history errors are those we've defined as neither
+// retryable nor worth sending to the DLQ.
+// TODO: We need a better name for this. ImpossibleToFixError? IsTerminalButNotDLQableError?
+func (e *executableImpl) isTerminalHistoryError(err error) (bool, error) {
 	if err == consts.ErrDependencyTaskNotCompleted {
 		metrics.TasksDependencyTaskNotCompleted.With(e.taggedMetricsHandler).Record(1)
-		return err
+		return true, err
 	}
 
 	if err == consts.ErrTaskRetry {
 		metrics.TaskStandbyRetryCounter.With(e.taggedMetricsHandler).Record(1)
-		return err
+		return true, err
 	}
 
 	if errors.Is(err, consts.ErrResourceExhaustedBusyWorkflow) {
 		metrics.TaskWorkflowBusyCounter.With(e.taggedMetricsHandler).Record(1)
-		return err
+		return true, err
```

Comment:
I am confused, why those are terminal errors? Task will be retried on those errors.

Reply:
Everything in this block was handled the same way, so I incorrectly assumed they were all discarded.

What's the contract for `HandleErr`? Are tasks discarded when we return `nil` rather than an error?

Reply:
Yes. Task will be acked/completed if nil is returned, otherwise task will keep retrying.